alns/accept/ThresholdAcceptance.py

```diff
@@ -0,0 +1,48 @@
+from alns.accept.RecordToRecordTravel import RecordToRecordTravel
+from alns.accept.update import update
+
+
+class ThresholdAcceptance(RecordToRecordTravel):
+    """
+    Threshold acceptance, using an updating threshold. The threshold is
+    updated as,
+
+    ``threshold = max(end_threshold, threshold - step)`` (linear)
+
+    ``threshold = max(end_threshold, step * threshold)`` (exponential)
+
+    where the initial threshold is set to ``start_threshold``.
+
+    Parameters
+    ----------
+    start_threshold
+        The initial threshold.
+    end_threshold
+        The final threshold.
+    step
+        The updating step.
+    method
+        The updating method, one of {'linear', 'exponential'}. Default
+        'linear'.
+
+    References
+    ----------
+    [1]: Santini, A., Ropke, S. & Hvattum, L.M. A comparison of acceptance
+         criteria for the adaptive large neighbourhood search metaheuristic.
+         *Journal of Heuristics* (2018) 24 (5): 783–815.
+    [2]: Dueck, G., Scheuer, T. Threshold accepting: A general purpose
+         optimization algorithm appearing superior to simulated annealing.
+         *Journal of Computational Physics* (1990) 90 (1): 161-175.
+    """
+
+    def __call__(self, rnd, best, current, candidate):
```

Comment:
Question:

This looks a lot like the implementation of RRT to me. What's different?

There's this 1993 paper about RRT, again by Dueck, where they also introduce great deluge. That's perhaps where Santini got his normalisation from, because the paper talks about some "record-deviation" parameter.

Reply:
RRT compares `candidate` against `best`, whereas TA compares it against `current`.

- I noticated that the paper by Dueck and Scheuer (1990) actually introduced TA. The paper in by Dueck in (1993) introduces RRT. 
- The "record-deviation" parameter in Dueck (1993) is a threshold value - similar to what the current implementation does. 
- Santini describes in the paper also that a "threshold starts at a large value and decreases at every iteration", which, to me, seems like his intention was to implement it using an absolute gap threshold. 
- If I had to take a guess based on my own experiences, I usually infer the absolute gap thresholds from the relative deviation, i.e., the initial threshold is 5% of the initial objective value. So directly implementing a threshold based on relative gap makes actually quite a lot of sense from a user POV.

Reply:
Good idea!